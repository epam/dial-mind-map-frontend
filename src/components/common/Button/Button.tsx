import classNames from 'classnames';
import { twMerge } from 'tailwind-merge';

interface ButtonProps {
  /** HTML type of the button (submit | reset | button) */
  htmlType?: 'submit' | 'reset' | 'button';
  /** Additional class names for the button */
  className?: string;
  /** Disabled state of the button */
  disabled?: boolean;
  /** Optional icon to display */
  icon?: React.ReactNode;
  /** Button text */
  label?: string;
  /** Click event handler */
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /**
   * Variant of the button, which determines additional styles:
   * - "primary" adds the "button-primary" classes,
   * - "secondary" adds the "button-secondary" classes,
   * - "chat" adds the "button-chat" classes.
   */
  variant?: 'primary' | 'secondary' | 'chat';
  /** Disable border even when variant is set */
  noBorder?: boolean;
  /** Position of the icon relative to the label */
  iconPosition?: 'before' | 'after';
  /** data-testid attribute for testing */
  dataTestId?: string;
}

const Button: React.FC<ButtonProps> = ({
  htmlType = 'button',
  className,
  disabled = false,
  icon,
  label,
  onClick,
  variant,
  noBorder = false,
  iconPosition = 'before',
  dataTestId,
}) => {
  const baseClasses = classNames('button', { [`button-${variant}`]: variant }, className);

  const btnClass = twMerge(baseClasses, noBorder && 'border-0');

  return (
    <button type={htmlType} disabled={disabled} onClick={onClick} className={btnClass} data-testid={dataTestId}>
      {icon && iconPosition === 'before' && <span className={classNames({ 'mr-2': !!label })}>{icon}</span>}

      {label && <span>{label}</span>}

      {icon && iconPosition === 'after' && <span className={classNames({ 'ml-2': !!label })}>{icon}</span>}
    </button>
  );
};

export default Button;
