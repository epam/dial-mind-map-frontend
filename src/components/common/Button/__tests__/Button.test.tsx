import { fireEvent, render, screen } from '@testing-library/react';

import Button from '../Button';

describe('Button component', () => {
  test('renders with default props', () => {
    render(<Button label="Click me" dataTestId="button" />);
    const button = screen.getByTestId('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('Click me');
    expect(button).toHaveAttribute('type', 'button');
  });

  test('calls onClick when clicked', () => {
    const handleClick = jest.fn();
    render(<Button label="Click me" onClick={handleClick} dataTestId="button" />);
    const button = screen.getByTestId('button');
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('does not call onClick when disabled', () => {
    const handleClick = jest.fn();
    render(<Button label="Disabled" disabled onClick={handleClick} dataTestId="button" />);
    const button = screen.getByTestId('button');
    fireEvent.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });

  test('renders with icon and default position (before)', () => {
    const Icon = () => <svg data-testid="icon" />;
    render(<Button icon={<Icon />} label="With Icon" dataTestId="button" />);
    const icon = screen.getByTestId('icon');
    // wrapper span should have margin-right when before
    expect(icon.parentElement).toHaveClass('mr-2');
    expect(screen.getByText('With Icon')).toBeInTheDocument();
  });

  test('positions icon after text when iconPosition="after"', () => {
    const Icon = () => <svg data-testid="icon" />;
    render(<Button icon={<Icon />} label="After Icon" iconPosition="after" dataTestId="button" />);
    const icon = screen.getByTestId('icon');
    // wrapper span should have margin-left when after
    expect(icon.parentElement).toHaveClass('ml-2');
  });

  test('applies variant classes correctly', () => {
    const { rerender } = render(<Button label="Primary" variant="primary" dataTestId="button" />);
    expect(screen.getByTestId('button')).toHaveClass('button-primary');

    rerender(<Button label="Secondary" variant="secondary" dataTestId="button" />);
    expect(screen.getByTestId('button')).toHaveClass('button-secondary');

    rerender(<Button label="Chat" variant="chat" dataTestId="button" />);
    expect(screen.getByTestId('button')).toHaveClass('button-chat');
  });

  test('applies noBorder correctly', () => {
    render(<Button label="No Border" variant="primary" noBorder dataTestId="button" />);
    const button = screen.getByTestId('button');
    expect(button).toHaveClass('border-0');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <Button
        label="Snapshot"
        icon={<span>*</span>}
        variant="primary"
        className="custom-class"
        noBorder
        iconPosition="after"
        dataTestId="button"
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
