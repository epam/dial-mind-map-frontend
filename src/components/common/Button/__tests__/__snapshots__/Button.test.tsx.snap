// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button-primary custom-class border-0"
    data-testid="button"
    type="button"
  >
    <span>
      Snapshot
    </span>
    <span
      class="ml-2"
    >
      <span>
        *
      </span>
    </span>
  </button>
</DocumentFragment>
`;
